concordium-consensus/src/Concordium/External.hs

```diff
@@ -298,6 +302,22 @@ handleStartExceptions logM c =
   where
     handleGlobalStateInitException (InvalidGenesisData _) = return (toStartResult StartGenesisFailure)
 
+-- |Migrate a legacy global state, if necessary.
+migrateGlobalState :: FilePath -> LogMethod IO -> IO ()
+migrateGlobalState dbPath logM = do
+    blockStateExists <- doesPathExist $ dbPath </> "blockstate-0" <.> "dat"
+    treeStateExists <- doesPathExist $ dbPath </> "treestate-0"
+    -- Only attempt migration when neither state exists
+    unless (blockStateExists || treeStateExists) $ do
+        oldBlockStateExists <- doesFileExist $ dbPath </> "blockstate" <.> "dat"
+        oldTreeStateExists <- doesDirectoryExist $ dbPath </> "treestate"
+        when (oldBlockStateExists && oldTreeStateExists) $ do
+            logM GlobalState LLInfo "Migrating global state from legacy version."
+            renameFile (dbPath </> "blockstate" <.> "dat") (dbPath </> "blockstate-0" <.> "dat")
+            renameDirectory (dbPath </> "treestate") (dbPath </> "treestate-0")
+            runLoggerT (addDatabaseVersion (dbPath </> "treestate-0")) logM
+            logM GlobalState LLInfo "Migration complete."
+
```

Comment:
Wouldn't it be cleaner if this also logged something if there is a partial database.

Additionally, we already have database versioning at the "outer layer", the database directory is passed in when the node is started, and it is currently `database-v4`. It seems confusing to have this versioning at two different layers in different naming schemes.

Reply:
Well it was doing versioning, but only one version was ever supported, and no migrations were done.

I think we should rename it to `database` only. I think the rust driver should control where the database location is, since some other data is also stored in the data directory, and we don't want accidental conflicts.

Reply:
Actually forget about that. It is easier if it just stays database-v4, otherwise migration is going to be problematic.

I would prefer that versioning would be done on that level, but with the split between R and H it is too much effort.